Initial

aljones15 opened the pull request:
HTTP APIs for [bedrock-payment](https://github.com/digitalbazaar/bedrock-payment/tree/initial)

aljones15 committed:
Add initial files.

aljones15 committed:
Add peerDeps use bedrock-payment.

aljones15 committed:
Update author.

aljones15 committed:
Add license.

aljones15 committed:
Add README.

aljones15 committed:
Use the orderProcessor from bedrock_payment.

aljones15 committed:
Use PUT for process payment.

aljones15 committed:
Check payment creator before setting PROCESSING.

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+const bedrock = require('bedrock');
```
missing copyright

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+const bedrock = require('bedrock');
+const brPassport = require('bedrock-passport');
+const {validate} = require('bedrock-validation');
+const asyncHandler = require('express-async-handler');
```
`asyncHandler` is exported by bedrock-express: https://github.com/digitalbazaar/bedrock-express/blob/master/lib/index.js#L22

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+      const {service} = req.query;
+      if(!service) {
+        throw new BedrockError(
+          'You must provide a service in the query', Errors.Data);
```
APIs called with missing/incorrect parameters is usually a `SyntaxError`.

aljones15 replied:
Ok went with that here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+        throw new BedrockError(
+          'You must provide a service in the query', Errors.Data);
+      }
+      logger.info(`Getting ${service} credentials`);
```
Use logging sparingly.  Is there a reason why we would want to log every hit to this API?  Usually, we just want to know if there's a problem, and bedrock-express takes care of logging if we just throw a BedrockError that includes debugging details.

aljones15 replied:
This is actually code from the last revision so I would imagine it was put there for debugging. I'll remove it.

aljones15 replied:
Removed it in this commit: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+      const data = await paymentService.getGatewayCredentials({service});
+      if(!data) {
+        throw new BedrockError(
+          `No Credentials Found for ${service}`, Errors.Data);
```
`NotFoundError` and you'd want to explicitly set the status code to 404.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+    PAYMENT_ENDPOINT,
+    ensureAuthenticated,
+    asyncHandler(async (req, res) => {
+      // this is so a user can not request
```
this comment seems confusing.  Perhaps just change the description of the API that it gets payments associated with the *authenticated* account.

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+    asyncHandler(async (req, res) => {
+      const {id: creator} = req.user.account;
+      const {payment: paymentData} = req.body;
+      logger.debug('CREATE PAYMENT CALLED', {paymentData});
```
remove non-essential logging.

aljones15 replied:
this might be useful in the refactor of bedrock-payment so I'm leaving it in. issue here: https://github.com/digitalbazaar/bedrock-payment-http/issues/2

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+          {creator, paymentData});
+        return res.status(201).json({order, payment});
+      } catch(e) {
+        logger.error('FAILED TO CREATE PAYPAL ORDER ERROR', {error: e});
```
Throw a BedrockError including the orignal error as the `cause` (last parameter)

aljones15 replied:
Take a look at the FIXME above it. This will be removed once the revisions to bedrock-payment are made so it work more atomically. issue here: https://github.com/digitalbazaar/bedrock-payment-http/issues/2

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+        return res.status(200).json({order: updatedOrder, payment});
+      }
+      logger.debug('CREATING NEW PAYMENT', {creator, paymentData});
+      // FIXME remove this try/catch before PR or leave a comment.
```
Address this?

aljones15 replied:
After Longley refactors bedrock-payment I will remove it. I'm making an issue to ensure these logs are removed: https://github.com/digitalbazaar/bedrock-payment-http/issues/2

gannan08 replied:
This `FIXME` talks about removing a try/catch... but I am failing to find one in this request handler.

aljones15 replied:
it was removed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/092fc2c5a7a8331f0151df314ccbe528d4329cd3

Thanks for catching this.

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+          updatePendingPayment({pendingPayments, paymentData});
+        return res.status(200).json({order: updatedOrder, payment});
+      }
+      logger.debug('CREATING NEW PAYMENT', {creator, paymentData});
```
If this logging should remain, make it into a properly cased sentence with punctuation. Consider any security implications of this data being in the log, anything sensitive in there?
```suggestion
      logger.debug('Creating a new payment.', {creator, paymentData});
```

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+    ensureAuthenticated,
+    validate({body: 'payment.processing'}),
+    asyncHandler(async (req, res) => {
+      logger.debug('PROCESS PAYMENT CALLED', {body: req.body});
```
remove logging.

aljones15 replied:
this might be useful when refactoring bedrock-payment so leaving it in. I will make it less SHOUTY CASE though.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+      const creator = account;
+      const {payment: paymentData, order = null} = req.body;
+      // get the payment
+      const payment = await paymentService.db.findOne(
```
can the `creator` be added to this query to prevent ever retrieving a payment for the wrong user, and therefore we don't have to check for that error condition later.

aljones15 replied:
This is a good idea, and I implemented it, but it does result in a more vague 404. That might be better as the account trying to use another account's payment gets a 404 instead of a 401. Less information for the attacker.

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+        throw new BedrockError(
+          'Payment not found',
+          Errors.NotFound,
+          {httpStatusCode: 400, public: true});
```
404 for NotFoundError.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+      // this throws a BedrockError if the payment is already finished.
+      // It has to be called before we make any change to the payment status.
+      paymentService.finished({payment});
+      if(payment.creator !== creator) {
```
I think this can be eliminated, see comment above.

aljones15 replied:
Good idea and did that here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,148 @@
+          Errors.NotAllowed, {httpStatusCode: 401, public: true});
+      }
+      payment.status = PaymentStatus.PROCESSING;
+      paymentService.db.save({payment});
```
this looks like an async function and probably needs an await?

aljones15 replied:
Great catch and thanks. Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/schemas/payment.js:
```diff
@@ -0,0 +1,93 @@
+const {config} = require('bedrock');
```
missing copyright.

mattcollier commented on lib/schemas/payment.js:
```diff
@@ -0,0 +1,93 @@
+const {config} = require('bedrock');
+const {PaymentStatus} = require('bedrock-payment');
+
+const w3cDateTime = require('bedrock-validation/schemas/w3cDateTime');
```
Don't access schemas through module loader, use the bedrock-validation schemas API. https://github.com/digitalbazaar/bedrock-validation/blob/master/lib/index.js#L21

Renaming a file internal to a module, in this case, bedrock-validation, should not be a breaking change if the public API remains unchanged.

See: https://github.com/digitalbazaar/bedrock-ledger-node/blob/master/schemas/webledger-validator.js#L7

aljones15 replied:
Addressed (hopefully correctly) here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on lib/schemas/payment.js:
```diff
@@ -0,0 +1,93 @@
+const {PaymentStatus} = require('bedrock-payment');
+
+const w3cDateTime = require('bedrock-validation/schemas/w3cDateTime');
+const numericFloat = '^[0-9]+(\\.[0-9]+)?$';
```
until we start dealing with wonky crytocurrency, seems like this should limit to two places behind decimal point.  Also, it appears to me that the decimal point is optional here, is that what is required?

aljones15 replied:
Decimal point is optional.

Also according this standard ISO_4217: http://travel.fyicenter.com/Currency/Currency_Decimal_Places_ISO_4217.php?P=3

some currencies allow 3 decimal places.

I'm not sure limiting the number of decimal places is a good idea as ISO conventions aside BigNumber should be able to handle a string with any number of decimals places.

mattcollier replied:
Do the interactions with Paypal use ISO_4217?

If an order is sent to paypal for US for '4.5555', does it work or blow up?

aljones15 replied:
GOOD CATCH SIR:

```
 "field": "/purchase_units/0/amount/value",
            "value": "14.555",
            "issue": "DECIMAL_PRECISION",
            "description": "If the currency supports decimals, only two decimal place precision is supported."
```

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/a25c3d8ebb2119f8cb6030a615f6478cabe8e426

mattcollier replied:
I plugged this regex in here and it doesn't seem to work, maybe I'm doing something wrong?  https://regex101.com/r/hhMydW/1

It also appears that the decimal point and digits afterward are still optional, does this mean that the paypal API in particular accepts integer values for USD amounts without difficulty?

aljones15 replied:
You left in the double escape for `.`

should be: `^[0-9]+(\.[0-9]{0,2})?$`
you have to double escape \\. for the json schema.

WRT to PayPal: yes.
PayPal will accept 5, 5.1, 5.25 but errors if you pass in 5.251.
According to their site on currencies some currencies do not accept decimals at all.

mattcollier commented on lib/schemas/payment.js:
```diff
@@ -0,0 +1,93 @@
+    },
+    currency: {
+      title: 'currency',
+      type: 'string'
```
Add `enum` here.

aljones15 replied:
Ok I used this: https://github.com/datasets/currency-codes/blob/master/data/codes-all.csv
to make a json array of all the currency codes and added it an enum. Good idea. Was a bit of a new feature and might have been better as a feature request, but was added here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/efae44a581e765338c880ee2f6b5e8d09f8cd39c

mattcollier replied:
We're accepting all these currencies by way of paypal?

aljones15 replied:
no we are definitely not accepting all of these currencies, but this validator is for Payments in general not for PayPal specifically.

mattcollier replied:
OK, at what point will we be validating against the list of currencies that we'll actually be accepting?

aljones15 replied:
@mattcollier Probably in the PayPal layer. I made an issue here: https://github.com/digitalbazaar/bedrock-payment-paypal/issues/1

Stripe did not accept Thai baht last time I checked, but PayPal does. Accepted currencies is definitely something to be checked in the Payment plugin. All we need to do here is make sure it's a valid currency at all.

aljones15 replied:
@mattcollier I believe we are happy with the validation here and in bedrock-payment-paypal correct?

mattcollier commented on lib/schemas/payment.js:
```diff
@@ -0,0 +1,93 @@
+    orders: {
+      title: 'orders',
+      type: 'array',
+      minItems: 1
```
Is there a schema for the array `items` here?

aljones15 replied:
no this is intentionally left open so we can support a variety of order types.

gannan08 replied:
What are some possible order types? Do we need to handle different order types with different logic paths?

aljones15 replied:
Possible order types:

1. Pay for a plan once.
2. Subscriptions for a Plan.
3. Charge when needed i.e. an account's balance is exhausted so they automatically recharge.
4. Whatever veres-wallet will be using this for.

The orders array is passed to the order processor. I believe the major concern with orders is simply just the size of it in the database. More than open to a discussion on what this data structure should be. At the moment for the accelerator it holds an array with a string for the single plan you purchase.

gannan08 replied:
It feels like there is a general data model we can deduce from this. We should find time to sort out what that is and ensure we fold that into this library. After, we should update the dependent libraries as well. We can plan time to work on this over voice and schedule something over chat.

aljones15 replied:
@gannan08 this branch now contains the updated data model. Validators have been updated to: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/8fca7d975f4cc829af1f6a28222b6ca7b75405d2

mattcollier commented on lib/schemas/payment.js:
```diff
@@ -0,0 +1,93 @@
+
+const processing = () => ({
+  type: 'object',
+  required: ['payment'],
```
if `order` is intentionally omitted here, please add a NOTE about why that is.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on package.json:
```diff
@@ -0,0 +1,45 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/digitalbazaar/bedrock-payment-http.git"
```
make https URL

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier commented on package.json:
```diff
@@ -0,0 +1,45 @@
+    "bedrock-payment": "0.0.1",
+    "bedrock-passport": "^5.0.0",
+    "bedrock-validation": "^4.1.0",
+    "express-async-handler": "^1.1.4"
```
this can be eliminated by using the export provided by bedrock-express.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/3dbb7eb8ce3be7e2d9e6d23ef1efbb2c6a1b31a1

mattcollier reviewed (changes requested):
Thank you!

aljones15 committed:
PR Fixes.

  - Add copyright to lib/index.js
  - Use bedrock-express asyncHandler
  - Use a syntax error if missing an query parameter.
  - Return 404 if credentials not found.
  - Improve GET /payment documentation.
  - REMOVE SHOUTY CASE LOGGING.
  - await save.
  - use bedrock-validation schemas.
  - Add copyright to lib/schemas/payment.js
  - Remove git+ from repo info.

aljones15 committed:
Restrict numericFloat to 2 decimal places.

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,151 @@
+      const {service} = req.query;
+      if(!service) {
+        throw new BedrockError(
+          'You must provide a service in the query', Errors.Syntax);
```
```suggestion
          'You must provide a service in the query.', Errors.Syntax);
```

mattcollier commented on lib/index.js:
```diff
@@ -0,0 +1,151 @@
+        throw new BedrockError(
+          'You must provide a service in the query', Errors.Syntax);
+      }
+      const data = await paymentService.getGatewayCredentials({service});
```
Seems like this API should throw a `NotFoundError` instead of returning undefined/null as shown here.

aljones15 replied:
That makes the code less re-usable by other services and it also hard codes an error when maybe we don't want to throw one. I actually really dislike this pattern of throwing on NotFound and it causes problems for me earlier because the balance service throws an error when a user doesn't have a balance yet. I guess my point is we can't guarantee that in every situation a payment gateway not having credentials is really an error. So I would prefer to keep the errors inside of the handler and separate them from function.

mattcollier replied:
There is a well established convention for `get` APIs to throw NotFoundError.  If you want to provide special handing for a NotFoundError you simply surround the API call with a try/catch and test `error.name === 'NotFoundError'`.

I don't care for APIs that required you to test for the very existence of a return value.  Also, this goes against using destructed assignment.  I like the contract with the API to be such that I can do the following without fear of a destructured assignment error when the API returns `undefined`.
```
const {foo} = await bar.getSomething();
```

aljones15 replied:
Can we agree to disagree here? I think programming inside of try/catch loops is horrible style and kind of defeats the purpose of try/catch. Catch, for me at least, is where you log an error and admit defeat, not where you add code to handle a situation you were expecting.

mattcollier replied:
throwing errors is the standard pattern for error handling with Promises and `try/catch` is the standard pattern for error handling with `async/await`.

Is it the case that @dlongley is providing the paymentService API, if so, we can see what he comes up with.

aljones15 replied:
A use case you might not have thought of is a user requests multiple credentials. In this case `null` would mean that one of those payment gateway credentials would fail, but another payment credential might be valid. I guess I just feel like throwing inside of this function hems it in to strongly and to soon.

aljones15 replied:
I might add on other projects I was expecting `null` and got an error thrown. What the absence of something means is really defined by the parent context. In this case the handler says `null` means an error in another it could mean something else.

aljones15 replied:
Addressed in https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/bd4395f747d33301e80694d02b24d0b7010cf706

aljones15 committed:
Add a period to the end of a comment.

Co-Authored-By: mattcollier <902041+mattcollier@users.noreply.github.com>

aljones15 committed:
Add a currency enum.

mattcollier commented:
@aljones15 a reminder... please do not force push during the PR process as comments can get lost/confused.

aljones15 committed:
Move throw to in GET /credentials into gateway plugin.

aljones15 committed:
Void payment is error has void in details.

aljones15 committed:
Errors are thrown in plugin process now.

aljones15 committed:
Update validator data model.

aljones15 committed:
Update parameters for updatePendingPayment.

aljones15 committed:
Remove try catch from create.

gannan08 commented on CHANGELOG.md:
```diff
@@ -0,0 +1,9 @@
+### Added
+  - lib/schemas/payment.js with validators for processing and creating payments.
+  - lib/index.js which sets up the configurable payment endpoints.
+  - Adds a route GET basePath/credentials for payment api keys / auth.
```
We can do without the "basePath/" in the changelog.

aljones15 replied:
I can remove it but the basePath is actually configurable. You can change it.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/092fc2c5a7a8331f0151df314ccbe528d4329cd3

gannan08 commented on README.md:
```diff
@@ -0,0 +1,54 @@
+<h1 align="center">Welcome to bedrock-payment-http 👋</h1>
```
Can we clean up this README to match our other projects?

aljones15 replied:
it is now a standard readme: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/092fc2c5a7a8331f0151df314ccbe528d4329cd3

gannan08 commented on lib/index.js:
```diff
@@ -0,0 +1,132 @@
+  const cfg = config['payment-http'];
+  const {basePath: PAYMENT_ENDPOINT} = cfg.routes;
+  // GET /payment/credentials
+  // GET the api keys/ auth tokens necessary to implement charges client side.
```
```suggestion
  // GET the api keys/auth tokens necessary to implement charges client side.
```

gannan08 commented on lib/schemas/currencies.json:
```diff
@@ -0,0 +1,304 @@
+[
```
Ideally, this should be its own module. I can see other libraries needing to pull in currency codes. It is fine here for now.

gannan08 commented on lib/schemas/payment.js:
```diff
@@ -0,0 +1,108 @@
+const {schemas} = require('bedrock-validation');
+const currencyEnum = require('./currencies.json');
+
+const numericFloat = '^[0-9]+(\\.[0-9]{0,2})?$';
```
Why are we doing this?

aljones15 replied:
currency is a string so we need to pass it in as a string and then use this regex (it comes from another project) to validate it. The regex has to be a string itself in order to work in the json schema validators so this means we have to double escape the `.` so regex does not mistake it for the regex `.` operator that matches anything I believe. Not 100% on what `.` means in a javascript RegEx.

Looked it up here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions


> . | (The decimal point) matches any single character except the newline character, by default.For example, /.n/ matches 'an' and 'on' in "nay, an apple is on the tree", but not 'nay'.If the s ("dotAll") flag is set to true, it also matches newline characters.
> -- | --

aljones15 committed:
Remove space from doc string.

Co-Authored-By: Ganesh Annan <gannan08@users.noreply.github.com>

aljones15 committed:
Update README to standard readme spec.

  - Update changelog.
  - Remove outdated FIXME.

aljones15 committed:
Add description string.

davidlehn commented on CHANGELOG.md:
```diff
@@ -0,0 +1,10 @@
+## 0.0.1 -
+
+### Added
+  - lib/schemas/payment.js with validators for processing and creating payments.
```
No need to indent these.

aljones15 replied:
fixed: https://github.com/digitalbazaar/bedrock-payment-http/pull/1/commits/15b56f878356ae8fe87cb90dee66017a31345754

aljones15 committed:
Remove indent from change log.

aljones15 committed:
Add full License details.
